x/bsonx/bsoncore/array.go

```diff
@@ -28,19 +28,20 @@ func NewArrayFromReader(r io.Reader) (Array, error) {
 	return newBufferFromReader(r)
 }
 
-// Index searches for and retrieves the element at the given index. This method will panic if
+// Index searches for and retrieves the value at the given index. This method will panic if
 // the array is invalid or if the index is out of bounds.
-func (a Array) Index(index uint) Element {
-	elem, err := indexErr(a, index)
+func (a Array) Index(index uint) Value {
+	value, err := a.IndexErr(index)
 	if err != nil {
 		panic(err)
 	}
-	return elem
+	return value
 }
 
-// IndexErr searches for and retrieves the element at the given index.
-func (a Array) IndexErr(index uint) (Element, error) {
-	return indexErr(a, index)
+// IndexErr searches for and retrieves the value at the given index.
+func (a Array) IndexErr(index uint) (Value, error) {
+	elem, err := indexErr(a, index)
+	return elem.Value(), err
```

Comment:
Looking at `indexErr`, there are cases where it may return `(nil, error)`, so these might not be valid. I don't think this would actually cause an issue because `Element#ValueErr` internally checks the element's length, which would be 0 for `nil`, but IMO idiomatic code shouldn't try to access the result if there was an error unless the API intentionally allows for this to happen. I'd change this to

```
elem, err := ...
if err != nil {
    return Value{}, err
}
return elem.Value(), nil
```

Reply:
That seems fair! Looks like `ValueErr()` does return `Value{}, ErrElementMissingType` if length is 0, but no point in touching `elem` if err is non-nil.